### PR TITLE
[QInfo broadcasting (2/3)] Support remaining functions in `qml.math.quantum`

### PR DESCRIPTION
**Pull request by eddddddy**
Currently the quantum information functions in `qml.math` and the QNode transforms in `qml.qinfo` do not support broadcasting. Support for this will be added in a sequence of 3 PRs, of which this is the second. The individual aims of the PRs are roughly:

1. (done) Split `qml.math.reduced_dm` into two functions, `qml.math.reduced_dm_from_sv` and `qml.math.reduced_dm_from_dm`, each of which supports broadcasting. Also begin a deprecation cycle for `qml.math.reduced_dm`.
2. (current) Support broadcasting for each of the remaining quantum info functions in `qml.math`. Also change their signatures to accept only a density matrix, and begin a deprecation cycle for passing in state vectors.
3. Support broadcasting for the `qml.qinfo` QNode transforms.

### CODE REVIEW

File: pennylane/math/multi_dispatch.py [hunk 1]
```diff
@@ -537,6 +537,11 @@ def einsum(indices, *operands, like=None):
     if like is None:
         like = get_interface(*operands)
     operands = np.coerce(operands, like=like)
+
+    if optimize is not None and like != "torch":
+        # torch einsum doesn't support the optimize keyword argument
+        return np.einsum(indices, *operands, like=like, optimize=optimize)
+
     return np.einsum(indices, *operands, like=like)
```
Comment (albi3ro) [inline pennylane/math/multi_dispatch.py:545]: Might be easier to read if we flip the logic. Controlling on the the negative forces the brain to jump through an extra hoop.

```suggestion
    if optimize is None or like == "torch":
        # torch einsum doesn't support the optimize keyword argument
        return np.einsum(indices, *operands, like=like)
    return np.einsum(indices, *operands, like=like, optimize=optimize)
```

File: pennylane/math/quantum.py [hunk 2]
```diff
@@ -449,7 +501,9 @@ def reduce_statevector(state, indices, check_state=False, c_dtype="complex128"):
     target = "".join(
         [ABC[i + 1] for i in sorted(indices)] + [ABC[num_wires + i + 1] for i in sorted(indices)]
     )
-    density_matrix = einsum(f"a{indices1},a{indices2}->a{target}", state, np.conj(state))
+    density_matrix = einsum(
+        f"a{indices1},a{indices2}->a{target}", state, np.conj(state), optimize="greedy"
```
Comment (mudit2812) [inline pennylane/math/quantum.py:505]: I'm curious why einsum optimization was not being used before 🤔

**Review by mudit2812 — APPROVED** [review]
Thanks for the changes @eddddddy . Looks good to me :)

File: pennylane/math/quantum.py [hunk 3]
```diff
@@ -585,11 +671,14 @@ def purity(state, indices, check_state=False, c_dtype="complex128"):
         # in the return value, such that these properties are transferred to the result
         # as well. Adding 2 instead of 1 because state[0] could be -1 in some cases, and
         # converting to a real number for jax grad compatibility.
-        return np.real((state[0] + 2.0) / (state[0] + 2.0))
+        if len(np.shape(state)) == 1:
+            return np.real((state[0] + 2.0) / (state[0] + 2.0))
+
+        return np.real((state[0][0] + 2.0) / (state[0][0] + 2.0))
```
Comment (timmysilv) [inline pennylane/math/quantum.py:677]: do we not need to worry about other batch dims?

File: pennylane/math/quantum.py [hunk 4]
```diff
@@ -31,6 +31,58 @@
...
+    if not len(shape) == 2:
+        raise ValueError("The state is not a state vector or a density matrix.")
+
+    if shape[0] != shape[1]:
+        # 2D arrays with different dims is a batched state vector
+        _state_param_deprecation_warning(func_name)
+        return True
+
+    # 2D arrays with same dims can be either; here we opt for density matrices
+    _interpret_dm_warning(x, func_name)
```
Comment (timmysilv) [inline pennylane/math/quantum.py:82]: to confirm, we want to always raise this warning until the next release?

**Review by timmysilv — COMMENTED** [review]
looks awesome! just a few little things

File: pennylane/math/quantum.py (hunk 4 above)
Comment (eddddddy) [reply pennylane/math/quantum.py:82]: Yep, we'll want to keep this warning for next release. Then when we drop support for state-vectors we'll remove it.

File: pennylane/math/quantum.py (hunk 3 above)
Comment (eddddddy) [reply pennylane/math/quantum.py:677]: Good catch 😄. This should indeed be a 1D array of 1s.

**Review by timmysilv — APPROVED** [review]

**Pull request merged**